tests/ui/features/lib/TextEditorPage.php

```diff
@@ -0,0 +1,219 @@
+
+		if (is_null($element)) {
+			throw new ElementNotFoundException(
+				"could not find element with xpath '" . $xpath . "'"
+			);
+		}
+
+		$keys = preg_split('//u', $text, null, PREG_SPLIT_NO_EMPTY);
+		$keys[] = Key::ENTER;
+		$element->postValue(array('value' => $keys));
+	}
+
+	/**
+	 * create a text file with the given name.
+	 * If name is not given the default is used.
+	 * If $useDefaultFileExtension is true, then only the name is entered and the
+	 * file extension is the default given by the application.
+	 *
+	 * @param Session $session
+	 * @param string $name
+	 * @param boolean $useDefaultFileExtension
+	 * @return void
+	 */
+	public function createTextFile(
+		Session $session,
+		$name = null,
+		$useDefaultFileExtension = false
+	) {
+		$newFileFolderButton
+			= $this->find("xpath", $this->newFileFolderButtonXpath);
+
+		if ($newFileFolderButton === null) {
+			throw new ElementNotFoundException(
+				"could not find new file/folder button"
+			);
+		}
+
+		$newFileFolderButton->click();
+
+		$newTextFileButton = $this->find("xpath", $this->newTextFileButtonXpath);
+
+		if ($newTextFileButton === null) {
+			throw new ElementNotFoundException(
+				"could not find new text file button"
+			);
+		}
+
+		$newTextFileButton->click();
+
+		if (strlen($name)) {
+			if ($useDefaultFileExtension) {
+				$this->typeInFieldAndPressEnter(
+					$this->newTextFileNameInputXpath,
+					$name,
+					$session
+				);
+			} else {
+				try {
+					$this->fillField($this->newTextFileNameInputLabel, $name . "\n");
+				} catch (\WebDriver\Exception\NoSuchElement $e) {
+					// this seems to be a bug in MinkSelenium2Driver.
+					// used to work fine in 1.3.1 but now throws this exception
+					// actually all that we need does happen,
+					// so we just don't do anything
+				}
+			}
+		} else {
+			$this->typeInFieldAndPressEnter(
+				$this->newTextFileNameInputXpath,
+				'',
+				$session
+			);
+		}
+
+		$this->waitForAjaxCallsToStartAndFinish($session);
+	}
+
+	/**
+	 * finds the textarea field to use for editing a text file
+	 *
+	 * @throws ElementNotFoundException
+	 * @return \Behat\Mink\Element\NodeElement
+	 */
+	public function findTextFileEditField() {
+		$textField = $this->find(
+			"xpath", $this->textFileEditXpath
+		);
+		if ($textField === null) {
+			throw new ElementNotFoundException("could not find textarea field");
+		}
+		return $textField;
+	}
+
+	/**
+	 * type text into the text area
+	 *
+	 * @param string $text
+	 * @return void
+	 */
+	public function typeIntoTextFile($text) {
+		$textField = $this->findTextFileEditField();
+		$textField->setValue($text);
+	}
+
+	/**
+	 * get the content of the open text file
+	 *
+	 * @return array of lines of text
+	 */
+	public function textFileContent() {
+		$textLayer = $this->find("xpath", $this->textFileTextLayerXpath);
+		if ($textLayer === null) {
+			throw new ElementNotFoundException("could not find text layer");
+		}
+		$textLineElements = $textLayer->findAll(
```

Comment:
@individual-it this returns me an array of elements that are the lines of text, but every element appears twice in the array.
Can you have a look and see what I am doing wrong in the xpath selector or?

(it is really consistent, with the code below to choose the text from every 2nd element, the tests never fail.)

Reply:
I think its because of the xpath `$textFileLineXpath = ".//div[contains(@class,'ace_line')]";`
It selects two divs `div class="ace_line_group"` and `div class="ace_line"`
![image](https://user-images.githubusercontent.com/2425577/29958031-69926912-8f10-11e7-88ef-4c3dfaac5784.png)

what about changing this xpath to: `.//div[@class='ace_line']` then you will have only the inner div

Reply:
That worked thank you
Now I understand that "contains" is just doing a substr kind of thing. I was hoping it would check through the classes (=words) in the class list of an element and choose if the element "contains" that class.